### PR TITLE
Generate bash completions automatically

### PR DESCRIPTION
**Pull request by Rylan12**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

See #10223

Even though this is a draft, I wouldn't mind some feedback on the general approach.

This PR will add functionality to automatically generate shell completion files every time a user runs `brew update`.

At the moment, this only works for bash and isn't very intelligent. This PR is mainly just to show my work in progress and get feedback along the way.

Here's my vision for this. The files we provide in `completions/` should be as minimal as possible. These source out to the user-specific configurations in `completions/user/`. These are automatically generated each time the user runs `brew update`. This has the benefit of automatically creating completions for all external commands (whether or official or not) without requiring the tap maintainers to provide them.

These completions should also be context-specific. For example, if `--foo` and `--bar` are mutually exclusive, already having `--foo` in the command means that `--bar` should never be suggested. This is not yet the case.

Formula and cask names should be filled intelligently. This means two things. First, if `--formula` or `--cask` is passed, only suggest formulae or casks. Second, certain commands should only accept installed formulae or casks. These should also be auto-filled appropriately. I don't yet know the best way to do this. We can use `named :formula` in some cases to detect that a formula argument is needed, but this won't catch all cases.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
> This PR will add functionality to automatically generate shell completion files every time a user runs `brew update`.

I think it'd be better to have these generated by us with `brew man` (or a similar command) and committed. Viewing the diffs is useful and avoids users having something potentially error-prone run on updates when the output will be identical for them all. This may also warrant pulling more logic from `brew update-report.rb` if it makes sense, too.

> This has the benefit of automatically creating completions for all external commands (whether or official or not) without requiring the tap maintainers to provide them.

This is a nice idea but I'm not sure it warrants the complexity without user requests for it. I'd like to see this scoped to just official (i.e. in our manpage) Homebrew commands for a first attempt and perhaps grow to support more external commands once the prior approach has been working well and the code tested in production. 

> We can use `named :formula` in some cases to detect that a formula argument is needed, but this won't catch all cases.

Yeh, this may benefit from a new DSL for this.

**Comment by MikeMcQuaid** [comment]
This is really cool work so far though and I'm looking forward to seeing it improve further 😍

**Comment by Rylan12** [comment]
Thanks!

Makes sense to me about restricting to official commands. In the future, it may be nice to allow this to be used for external commands, but maybe we can, instead, provide an easy way for tap maintainers to provide their own completions (possible autogenerated and possibly not) and then link in `Tap.link_completions_and_manpages`.

In terms of DSL, I think I'll need to look more closely at what exactly our commands use and in which cases the `named` method is lacking sufficient information. One thing I was thinking about it using something like one of the following:

```ruby
named type: :formula
named type: :cask
named type: :formula_or_cask
named type: :installed_formula
named type: :installed_cask
named type: :installed_formula_or_cask
named type: :installed_formula, count: 1
named type: :installed_formula, min_count: 1
named type: :installed_formula, min_count: 1, max_count: 2
```

**Comment by Rylan12** [comment]
One question, though:

What's the best approach to replacing text in the completions files? Another benefit to the user completions directory is that the generated text is dumped to its own file so it didn't need to actually replace text in the completions file. Is having `brew man` write to a separate file that's sourced by the completion file still an okay way to do it or should `brew man` actually rewrite the completions file?

**Comment by MikeMcQuaid** [comment]
> Makes sense to me about restricting to official commands. In the future, it may be nice to allow this to be used for external commands, but maybe we can, instead, provide an easy way for tap maintainers to provide their own completions (possibly autogenerated and possibly not) and then link in `Tap.link_completions_and_manpages`.

Yeh, I think we do already do link those is available e.g. https://github.com/Homebrew/homebrew-services/tree/master/completions/bash

> In terms of DSL, I think I'll need to look more closely at what exactly our commands use and in which cases the `named` method is lacking sufficient information. One thing I was thinking about it using something like one of the following:

Yeh, those examples look nice. They could also be perhaps used to generate the `install <formula>` first lines too in the manpage/help output (to ensure that they are consistent).





> Is having `brew man` write to a separate file that's sourced by the completion file still an okay way to do it or should `brew man` actually rewrite the completion file?

I think it should write the entire completion file from scratch (for now at least) in the same way `brew man` does with the manpage. In future we could consider having a separate file which is auto included (but there's some risk run from having `brew tap` automatically add code to  your `.bashrc` which is run on new shells from an untrusted and auto-updated source...).

**Comment by Rylan12** [comment]
> > Makes sense to me about restricting to official commands. In the future, it may be nice to allow this to be used for external commands, but maybe we can, instead, provide an easy way for tap maintainers to provide their own completions (possibly autogenerated and possibly not) and then link in `Tap.link_completions_and_manpages`.
> 
> Yeh, I think we do already do link those is available e.g. https://github.com/Homebrew/homebrew-services/tree/master/completions/bash

Oh, nice! I should have guessed based on the word "completions" 😅

> > In terms of DSL, I think I'll need to look more closely at what exactly our commands use and in which cases the `named` method is lacking sufficient information. One thing I was thinking about it using something like one of the following:
> 
> Yeh, those examples look nice. They could also be perhaps used to generate the `install <formula>` first lines too in the manpage/help output (to ensure that they are consistent).

Good idea

> > Is having `brew man` write to a separate file that's sourced by the completion file still an okay way to do it or should `brew man` actually rewrite the completion file?
> 
> I think it should write the entire completion file from scratch (for now at least) in the same way `brew man` does with the manpage. In future we could consider having a separate file which is auto included (but there's some risk run from having `brew tap` automatically add code to your `.bashrc` which is run on new shells from an untrusted and auto-updated source...).

Okay, fair enough. Although, the current way I do it doesn't require writing to `.bashrc`. There's a `source` in the completions file that points to `$HOMEBREW_REPOSITORY/completions/wherever`.

One more thing: I'm going to remove all of the `brew cask <command>` completions as they're disabled anyway.

**Comment by MikeMcQuaid** [comment]
> There's a `source` in the completions file that points to `$HOMEBREW_REPOSITORY/completions/wherever`.

Yeh, my point is if we're updating that with no user intervention required and don't require any user intervention to opt-in from the current completions setup: we're going to lower the trust profile for that in a way I think is undesirable. Does that makes sense? If not I can explain with a better example.



> One more thing: I'm going to remove all of the `brew cask <command>` completions as they're disabled anyway.

🎉

**Comment by alebcay** [comment]
Currently, on `brew update-report`, we run `rebuild_commands_completion_list` which automatically adds user-specified commands into the completions list (e.g. those from third-party taps). That functionality was added in https://github.com/Homebrew/brew/pull/7767.

I think the code added in that PR may be worth refactoring, repurposing, or perhaps removing with the changes in this PR.

**Comment by Rylan12** [comment]
> > There's a `source` in the completions file that points to `$HOMEBREW_REPOSITORY/completions/wherever`.
> 
> Yeh, my point is if we're updating that with no user intervention required and don't require any user intervention to opt-in from the current completions setup: we're going to lower the trust profile for that in a way I think is undesirable. Does that makes sense? If not I can explain with a better example.

We already link the completions each time `brew update` is run, right? The user currently has to opt-in by adding the correct lines to their `.bashrc` file. Any `brew update` could theoretically put malicious code into the completion file which would be executed on the next shell startup, right? I understand the concern, I think, just not why it's any worse.

**Comment by MikeMcQuaid** [comment]
> I understand the concern, I think, just not why it's any worse.

Yeh, you're right, I'm just thinking we should/could consider rolling this back, I dunno.

**Comment by Rylan12** [comment]
> Currently, on `brew update-report`, we run `rebuild_commands_completion_list` which automatically adds user-specified commands into the completions list (e.g. those from third-party taps) - at least on `zsh`. That functionality was added in #7767.

Are you suggesting we don't autocomplete any user commands? If so, yeah, that can be deleted I think.

**Comment by Rylan12** [comment]
> Yeh, you're right, I'm just thinking we should/could consider rolling this back, I dunno.

What would a rollback look like? We still want to offer completions, I assume. Does that mean users will be required to manually update their completions on each update?

If so, we should make this as easy as possible I think. Preferably with a command or option to easily do it while still requiring users to actually take the action.

**Comment by alebcay** [comment]
> Are you suggesting we don't autocomplete any user commands? If so, yeah, that can be deleted I think.

I think we currently do, but the mechanism is different depending on the shell. I believe the current brew/fish completions invoke `brew commands` and get the info from there. The code presented in that PR creates a list that gets updated on every `brew update-report`. Thus the autocomplete can just read the list from a text file rather than invoking a full Ruby environment to get the list of available (internal and user) commands.

The hope was that we could leverage the cached commands list in the brew/fish completions so that they can also be faster but as far as I can tell those improvements have not been made yet - so far, only the zsh completions use it.

**Comment by Rylan12** [comment]
That is used for bash too, I believe:

https://github.com/Homebrew/brew/blob/dad7dc6a1498b80770d98f2d7cd6fb927c300bbb/completions/bash/brew#L96-L105

Either way, I agree that it should be used for fish as well. I don't use fish at all, but I'll see what I can do. Good to keep in mind for sure.

**Comment by MikeMcQuaid** [comment]
> Are you suggesting we don't autocomplete any user commands?

Yeh and also perhaps we don't link completions from non-Homebrew org taps automatically.

> The hope was that we could leverage the cached commands list in the brew/fish completions so that they can also be faster but as far as I can tell those improvements have not been made yet - so far, only the zsh completions use it.

Yup.

**Comment by Rylan12** [comment]
Okay, sounds good. Any preference for a method of enabling/disabling linking?

This probably deserves a separate PR, no? (I'd be happy to do it)

**Comment by MikeMcQuaid** [comment]
> Okay, sounds good. Any preference for a method of enabling/disabling linking?

I'm not sure! I think just anything that doesn't work automatically/by default. Could be an environment variable or just linking them into another location?

> This probably deserves a separate PR, no?

Agreed 👍🏻

**Comment by Rylan12** [comment]
Here are my tentative thoughts on the DSL (probably coming in a separate PR). Feedback is always appreciated.

Here's the method I'm proposing:

```ruby
def completions(named)
```

**`named`** defines the types of named arguments. The (tentative) options are:
- `none:` don't include completions for this command
- `no_named:` no named args (not needed when `options_only` or `exclude` are passed to `type`)
- `installed:` complete installed formulae and casks
- `formula:` complete only available formulae
- `cask:` complete only available casks
- `installed_formula:` complete only formulae that are installed
- `installed_cask:` complete only casks that are installed
- Passing an array of strings will complete those options (useful for e.g. `on`, `off` in `analytics`)

---

Some examples:

```ruby
# update-report
completions :none

# install
completions :formulae_and_casks

# uninstall
completions :installed

# link
completions :formula

# autoremove
completions :no_named

# analytics (one note: I'm not sure how to handle the global args here)
completions %w[state on off regenerate-uuid]
```

Of course, I'll refine this as I work through the commands to determine what exactly we need.

File: Library/Homebrew/completions.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/completions.rb:79]: Could put some of this into an `erb` file and inject only the dynamic bits?

**Review by MikeMcQuaid — COMMENTED** [review]
Looking good so far!

**Comment by MikeMcQuaid** [comment]
> * `none:` don't include completions for this command
> * `no_named:` no named args (not needed when `options_only` or `exclude` are passed to `type`)
> * `installed:` complete installed formulae and casks
> * `formula:` complete only available formulae
> * `cask:` complete only available casks
> * `installed_formula:` complete only formulae that are installed
> * `installed_cask:` complete only casks that are installed
> * Passing an array of strings will complete those options (useful for e.g. `on`, `off` in `analytics`)

These all look good but I wonder if this should be called `completions` because most/all of these relate to the command itself rather than just the completions. We could/should ideally use this to actually have the command-line parsing handle these cases for us rather than each command doing it manually.

File: Library/Homebrew/completions.rb
Comment (Rylan12) [reply Library/Homebrew/completions.rb:79]: Yes, this is exactly what I was looking for!

**Comment by Rylan12** [comment]
Is it worth repurposing/replacing the `named` method? Or is there a better way?

**Comment by MikeMcQuaid** [comment]
> Oh, `named` might not be a method we've written.

It is 👍🏻 

> Is it worth repurposing/replacing the `named` method? Or is there a better way?

Yeh, this seems fine to me!

> Maybe, instead, we can create our own method that will call `named` appropriately based on what's passed.


Also fine with me!

File: completions/bash/brew
Comment (max-ae) [inline completions/bash/brew:125]: This probably shouldn't get listed twice here.

Comment (Rylan12) [reply completions/bash/brew:125]: Good catch, thanks!

File: Library/Homebrew/cli/parser.rb
Comment (SeekingMeaning) [inline Library/Homebrew/cli/parser.rb:504]:
```suggestion
        @processed_options.reject! { |existing| existing.second == option.long.first } if option.long.first.present?
```

File: Library/Homebrew/commands.rb
Comment (SeekingMeaning) [inline Library/Homebrew/commands.rb:200]: @MikeMcQuaid what do you think about using comments like `# TODO (2.8):` for things that should be done on `brew` minor version updates?

File: Library/Homebrew/commands.rb
Comment (SeekingMeaning) [inline Library/Homebrew/commands.rb:210]:
```suggestion
    path = self.path(command)
    return if path.blank?
```

File: Library/Homebrew/cmd/options.rb
Comment (SeekingMeaning) [inline Library/Homebrew/cmd/options.rb:44]: Is `odie` still needed?

```rb
odie "Unknown command: #{args.command}" if cmd_options.nil?
```

Note: `[].nil?` returns `false`, but `[].blank?` returns `true`

File: Library/Homebrew/commands.rb
Comment (SeekingMeaning) [inline Library/Homebrew/commands.rb:224]: I know you didn't originally write this, but I didn't know that we could declare variables from regex(!)

File: Library/Homebrew/completions.rb
Comment (SeekingMeaning) [inline Library/Homebrew/completions.rb:20]: 💯

File: Library/Homebrew/completions/bash.erb
Comment (SeekingMeaning) [inline Library/Homebrew/completions/bash.erb:13]: Nice, maybe could even add this to a doc somewhere (if it hasn't already)

File: Library/Homebrew/commands.rb
Comment (Rylan12) [reply Library/Homebrew/commands.rb:200]: This is probably a good idea. Otherwise, it's a ton to keep track of. We may as well our part to help Mike out if he's the one doing the deprecations 😉

File: Library/Homebrew/cmd/options.rb
Comment (Rylan12) [reply Library/Homebrew/cmd/options.rb:44]: You're right that the `odie` should still be here. To clarify, you're suggesting this, right?

```ruby
    elsif args.command.present?
      cmd_options = Commands.command_options(args.command)
      odie "Unknown command: #{args.command}" if cmd_options.nil?
      ...
```

**Comment by Rylan12** [comment]
Thanks for the review, @SeekingMeaning! Good suggestions. Just so you know, I'm putting this on hold temporarily while I finish up implementing the DSL for the named arguments (which will block this PR). I'll implement your suggestions when I get back to this.

File: Library/Homebrew/cmd/options.rb
Comment (SeekingMeaning) [reply Library/Homebrew/cmd/options.rb:44]: Yep! 👍

File: Library/Homebrew/commands.rb
Comment (jonchang) [reply Library/Homebrew/commands.rb:200]: We already have a bunch of these in the codebase (grep for '# TODO'), so it's definitely a good idea.

Comment (MikeMcQuaid) [reply Library/Homebrew/commands.rb:200]: the easiest thing that gets me to find them on minor updates is a commented-out `odeprecated` or `odisabled` (because that's what I search for).

File: Library/Homebrew/completions/bash.erb
Comment (Rylan12) [reply Library/Homebrew/completions/bash.erb:13]: You know what, this information can be found (somewhat) in the [external commands documentation](https://docs.brew.sh/External-Commands). I also have a separate PR open to improve this. I'd say if there's something you think is missing, leave a comment/review over in https://github.com/Homebrew/brew/pull/10333 and I'll try to improve it there.

**Comment by Rylan12** [comment]
Alright, here's where we are at the moment:

Running `brew man` will automatically generate completions for all internal brew commands. These completions will complete all options that can be passed to the given command. The completions will also complete certain types of named arguments when defined using `named_args`.

Here are the types of named arguments and their `named_args` counterpart (they're pretty self-explanatory tbh):
- None (default or `:none`)
- Formulae (`:formula`)
- Installed formulae (`:installed_formula`)
- Outdated formulae (`:outdated_formula`) **New**
- Casks (`:cask`)
- Installed casks (`:installed_cask`)
- Outdated casks (`:outdated_cask`) **New**
- Installed taps (`:tap`, `:installed_tap`)
- Commands (`:command`)
- Diagnostic checks (`:diagnostic_check`)
- Custom subcommands e.g. `on`, `off`, `state`, `regenerate-uuid` for `brew analytics` (string array)

At the moment, commands that specify multiple types of arguments will only be able to complete the type that's defined last. For example, `brew uninstall` only completes casks. I think I need to learn a little more about how bash completions work to come up with the best solution.

**Comment by MikeMcQuaid** [comment]
> Additionally, these completions lack context-awareness. Meaning, it will suggest `--formula` even if `--cask` is passed. Based on the conversations here and in #10223, I think this is okay for the first round of this.

Definitely 👍🏻 



> At the moment, commands that specify multiple types of arguments will only be able to complete the type that's defined last. For example, `brew uninstall` only completes casks. I think I need to learn a little more about how bash completions work to come up with the best solution.

I wish I knew enough about completions to help here but agreed this is a blocker for now.

**Comment by Rylan12** [comment]
Okay, the completions should work as expected, now. I have to say, I feel much more confident with how our bash completions work now, so that's good.

For the record, here are the resources I used:
- https://iridakos.com/programming/2018/03/01/bash-programmable-completion-tutorial
- https://www.gnu.org/software/bash/manual/html_node/Programmable-Completion-Builtins.html#Programmable-Completion-Builtins
- https://www.gnu.org/software/bash/manual/html_node/Bash-Variables.html#Bash-Variables
- https://tylerthrailkill.com/2019-01-19/writing-bash-completion-script-with-subcommands/ (this was interesting because the way the author learned how to use completions was by reverse-engineering our completions 😄)

---

I think the main things missing that would make this extra great are:
- Only completing named arguments until the maximum is reached. For example, if a command accepts exactly one named argument, don't complete anything if an argument is already passed.
- Don't complete suggestions that conflict with each other (e.g. never suggest `--cask` if `--formula` was already passed)
- Don't suggest casks if `--formula` is passed and vice versa

These are a little more tricky, so I'm not planning on implementing them in this PR. I have some general ideas about how to do it, though, so hopefully, I'll be able to figure out a way to make it work with relative ease.

**Review by MikeMcQuaid — APPROVED** [review]
Incredible work here @Rylan12! Do you foresee yourself doing ZSH and/or Fish completions too?

**Comment by Rylan12** [comment]
> Incredible work here @Rylan12!

Thanks!

> Do you foresee yourself doing ZSH and/or Fish completions too?

Yes, my plan is to work on those next (probably zsh before fish). I'd rather do it in a separate PR, though, to keep things cleaner and easier to keep track of.

**Comment by MikeMcQuaid** [comment]
> Yes, my plan is to work on those next (probably zsh before fish). I'd rather do it in a separate PR, though, to keep things cleaner and easier to keep track of.

Agreed and: great!

**Pull request merged**
